File: sdk-node/test/unit/core/headers.test.ts
import { describe, expect, it } from 'vitest'

import { buildHeaders } from '@/core/headers'

describe('buildHeaders', () => {
  const base = {
    apiKey: '3co_test',
    apiVersion: '2026-04-29',
    sdkVersion: '0.0.0-test',
    userAgentSuffix: 'Node/v22.0.0; darwin-arm64-24.0.0',
    telemetryHeader: undefined,
    idempotencyKey: undefined,
  }

  it('sets the standard request headers', () => {
    const headers = buildHeaders(base)
    expect(headers.get('Authorization')).toBe('Bearer 3co_test')
    expect(headers.get('Threecommon-Version')).toBe('2026-04-29')
    expect(headers.get('User-Agent')).toBe(
      'ThreeCommonNode/0.0.0-test (Node/v22.0.0; darwin-arm64-24.0.0)',
    )
    expect(headers.get('Accept')).toBe('application/json')
    expect(headers.get('Content-Type')).toBe('application/json')
  })

  it('omits the telemetry header when value is undefined', () => {
    const headers = buildHeaders(base)
    expect(headers.has('Threecommon-Client-Telemetry')).toBe(false)
  })

  it('attaches the telemetry header when supplied', () => {
    const headers = buildHeaders({ ...base, telemetryHeader: '{"lang":"node"}' })
    expect(headers.get('Threecommon-Client-Telemetry')).toBe('{"lang":"node"}')
  })

  it('omits Idempotency-Key when value is undefined', () => {
    const headers = buildHeaders(base)
    expect(headers.has('Idempotency-Key')).toBe(false)
  })

  it('attaches Idempotency-Key when supplied', () => {
    const headers = buildHeaders({ ...base, idempotencyKey: 'abc-123' })
    expect(headers.get('Idempotency-Key')).toBe('abc-123')
  })
})

File: sdk-go/internal/core/headers_test.go
package core_test

import (
	"testing"

	"github.com/stretchr/testify/assert"

	"github.com/3-Common/sdk/sdk-go/internal/core"
)

func TestBuildHeaders_RequiredFieldsPopulated(t *testing.T) {
	t.Parallel()

	h := core.BuildHeaders(core.HeadersInput{
		APIKey:          "3co_test",
		APIVersion:      "2026-04-29",
		SDKVersion:      "0.1.0",
		UserAgentSuffix: "Go/go1.22; darwin-arm64",
		HasBody:         true,
	})

	assert.Equal(t, "Bearer 3co_test", h.Get("Authorization"))
	assert.Equal(t, "2026-04-29", h.Get("Threecommon-Version"))
	assert.Equal(t, "ThreeCommonGo/0.1.0 (Go/go1.22; darwin-arm64)", h.Get("User-Agent"))
	assert.Equal(t, "application/json", h.Get("Accept"))
	assert.Equal(t, "application/json", h.Get("Content-Type"))
}

func TestBuildHeaders_ContentTypeOmittedWhenBodyless(t *testing.T) {
	t.Parallel()

	h := core.BuildHeaders(core.HeadersInput{APIKey: "k", HasBody: false})
	assert.Empty(t, h.Get("Content-Type"))
}

func TestBuildHeaders_ContentTypeSetWhenBodyPresent(t *testing.T) {
	t.Parallel()

	h := core.BuildHeaders(core.HeadersInput{APIKey: "k", HasBody: true})
	assert.Equal(t, "application/json", h.Get("Content-Type"))
}

func TestBuildHeaders_OptionalFieldsOmittedWhenEmpty(t *testing.T) {
	t.Parallel()

	h := core.BuildHeaders(core.HeadersInput{APIKey: "k"})
	assert.Empty(t, h.Get("Threecommon-Client-Telemetry"))
	assert.Empty(t, h.Get("Idempotency-Key"))
}

func TestBuildHeaders_OptionalFieldsSetWhenProvided(t *testing.T) {
	t.Parallel()

	h := core.BuildHeaders(core.HeadersInput{
		APIKey:          "k",
		TelemetryHeader: `{"lang":"go"}`,
		IdempotencyKey:  "key-1",
	})
	assert.Equal(t, `{"lang":"go"}`, h.Get("Threecommon-Client-Telemetry"))
	assert.Equal(t, "key-1", h.Get("Idempotency-Key"))
}
